File: code/frontend/src/utils/diff.js
import { diff_match_patch } from 'diff-match-patch'

function encode(str) {
  return encodeURIComponent(str.replace(/%20/g, '+'))
}

function unescaper(delta) {
  return delta
    .replace(/%21/g, '!')
    .replace(/%7E/g, '~')
    .replace(/%27/g, '\'')
    .replace(/%28/g, '(')
    .replace(/%29/g, ')')
    .replace(/%3B/g, ';')
    .replace(/%2F/g, '/')
    .replace(/%3F/g, '?')
    .replace(/%3A/g, ':')
    .replace(/%40/g, '@')
    .replace(/%26/g, '&')
    .replace(/%3D/g, '=')
    .replace(/%2B/g, '+')
    .replace(/%24/g, '$')
    .replace(/%2C/g, ',')
    .replace(/%23/g, '#')
    .replace(/%2A/g, '*')
}


diff_match_patch.prototype.diff_toDelta = function (diffs) {
  let delta = []

  for (let i = 0; i < diffs.length; i++) {
    const [op, text] = diffs[i]

    switch (op) {
    case diff_match_patch.DIFF_INSERT:
      const encodedInsert = encode(text).replace(/\+/g, " ") // eslint-disable-line
      delta.push(`+${encodedInsert}`)
      break
    case diff_match_patch.DIFF_DELETE:
        const deleteCount = [...text].length // eslint-disable-line
      // or Array.from(text).length
      delta.push(`-${deleteCount}`)
      break
    case diff_match_patch.DIFF_EQUAL:
      const equalCount = [...text].length // eslint-disable-line
      delta.push(`=${equalCount}`)
      break
    }
  }

  const deltaParts = delta.join('\t')
  return unescaper(deltaParts)
}

diff_match_patch.prototype.diff_fromDelta = function (text, delta) {
  let diffs = []
  let i = 0

  const runes = [...text]
  const tokens = delta.split('\t')

  for (let token of tokens) {
    if (token.length === 0) {
      continue
    }

    const op = token.charAt(0)
    let param = token.slice(1)

    if (op === '+') {
      param = param.replace(/\+/g, '%2b')
      try {
        param = decodeURIComponent(param)
      } catch (err) { // eslint-disable-line
        throw new Error(`Error decoding insert token: "${param}"`)
      }
      diffs.push([diff_match_patch.DIFF_INSERT, param])
    } else if (op === '-' || op === '=') {
      const n = parseInt(param, 10)
      if (isNaN(n) || n < 0) {
        throw new Error('Invalid number in DiffFromDelta:' + param)
      }
      i += n
      if (i > runes.length) {
        break
      }
      const text = runes.slice(i - n, i).join('')
      if (op === '-') {
        diffs.push([diff_match_patch.DIFF_DELETE, text])
      } else {
        diffs.push([diff_match_patch.DIFF_EQUAL, text])
      }
    } else {
      throw new Error('Unknown DiffOp: ' + op)
    }
  }
  if (i !== runes.length) {
    throw new Error('Delta length larger than source text')
  }

  return diffs
}

export const shouldCreateSnapshot = (versionNumber) => {
  return versionNumber % 10 === 0
}

export const decryptVersionContent = async (version, encryptionService, symmetricKey) => {
  try {
    const noteCipherKey = await encryptionService.unwrapNoteCipherKey(version.cipherKey, version.cipherIv, symmetricKey)
    return await encryptionService.decryptNoteContent(version.content, version.contentIv, noteCipherKey)
  } catch (error) {
    console.error('Failed to decrypt version content:', error)
    throw new Error('Failed to decrypt version content')
  }
}

export const createEncryptedDiff = async (baseContent, newContent, encryptionService, symmetricKey, noteKeys) => {
  const dmp = new diff_match_patch()

  const diffs = dmp.diff_main(baseContent, newContent, false)
  dmp.diff_cleanupEfficiency(diffs)
  const diffDelta = dmp.diff_toDelta(diffs)

  return await encryptVersionContent(diffDelta, encryptionService, symmetricKey, noteKeys)
}

export const buildVersionContent = async (chain, encryptionService, symmetricKey) => {
  if (!chain || chain.length === 0) return ''

  if (chain[0].type !== 'snapshot') {
    console.warn(`Warning: Chain doesn't start with snapshot (starts with ${chain[0].type})`)
    return ''
  }

  const dmp = new diff_match_patch()

  let content
  try {
    content = await decryptVersionContent(chain[0], encryptionService, symmetricKey)
  } catch (error) {
    console.error('Failed to decrypt initial snapshot:', error)
    throw new Error('Failed to decrypt initial snapshot')
  }

  for (let i = 1; i < chain.length; i++) {
    const version = chain[i]
    if (version.type !== 'diff') {
      console.warn(`Warning: Unexpected version type in chain: ${version.type}`)
      continue
    }

    try {
      const decryptedDiff = await decryptVersionContent(version, encryptionService, symmetricKey)

      const diffs = dmp.diff_fromDelta(content, decryptedDiff)
      const patches = dmp.patch_make(content, diffs)
      const [newContent, results] = dmp.patch_apply(patches, content)

      if (!results.every(result => result)) {
        console.warn(`Warning: Some patches failed to apply for version #${version.metadata.versionNumber}`)
        continue
      }

      content = newContent
    } catch (error) {
      console.error(`Warning: Failed to process version #${version.metadata.versionNumber}:`, error)
      continue
    }
  }

  return content
}

export const encryptVersionContent = async (content, encryptionService, symmetricKey, noteKeys) => {
  try {
    const noteCipherKey = await encryptionService.unwrapNoteCipherKey(noteKeys.cipherKey, noteKeys.cipherIv, symmetricKey)

    const { encryptedContent, iv: contentIv } = await encryptionService.encryptNoteContent(content, noteCipherKey)

    return {
      encryptedContent,
      contentIv,
      cipherKey: noteKeys.cipherKey,
      cipherIv: noteKeys.cipherIv
    }
  } catch (error) {
    console.error('Failed to encrypt version content:', error)
    throw new Error('Failed to encrypt version content')
  }
}


export default diff_match_patch

File: code/client/internal/utils/diff.go
package utils

import (
	"fmt"
	"github.com/sergi/go-diff/diffmatchpatch"
	"noted/internal/api"
	"noted/internal/encryption"
	"strings"
)

func SelectVersion(versions []*api.Version) (*api.Version, error) {
	fmt.Println("Available versions:")
	fmt.Println()

	var maxVerLen, maxTitleLen, maxTagsLen int
	for _, v := range versions {
		verLen := len(fmt.Sprintf("#%d", v.Metadata.VersionNumber))
		if verLen > maxVerLen {
			maxVerLen = verLen
		}
		if len(v.Metadata.Title) > maxTitleLen {
			maxTitleLen = len(v.Metadata.Title)
		}
		tagsLen := len(strings.Join(v.Metadata.Tags, ", "))
		if tagsLen == 0 {
			tagsLen = 4
		}
		if tagsLen > maxTagsLen {
			maxTagsLen = tagsLen
		}
	}

	if maxVerLen < 3 {
		maxVerLen = 3
	}

	headerFmt := fmt.Sprintf("%%-%ds  %%-20s  %%-%ds  %%-%ds\n", maxVerLen, maxTitleLen, maxTagsLen)
	fmt.Printf(headerFmt, "Ver", "Created", "Title", "Tags")
	fmt.Println(strings.Repeat("-", maxVerLen+maxTitleLen+maxTagsLen+27))

	length := len(versions)
	for i := length - 1; i >= 0; i-- {
		v := versions[i]
		tags := strings.Join(v.Metadata.Tags, ", ")
		if tags == "" {
			tags = "no tags"
		}

		fmt.Printf(fmt.Sprintf("%%-%ds  %%-20s  %%-%ds  %%-%ds\n", maxVerLen, maxTitleLen, maxTagsLen),
			fmt.Sprintf("#%d", v.Metadata.VersionNumber),
			v.CreatedAt.Format("2006-01-02 15:04:05"),
			v.Metadata.Title,
			tags,
		)
	}

	var choice int
	fmt.Print("\nChoose a number (1-", length, "): ")
	_, err := fmt.Scanf("%d", &choice)
	if err != nil || choice < 1 || choice > length {
		return nil, fmt.Errorf("invalid selection")
	}

	selectedIndex := length - choice
	return versions[selectedIndex], nil
}

func BuildDecryptedVersionContent(chain []*api.Version, encryptionService *encryption.EncryptionService, symmetricKey []byte) (string, error) {
	if len(chain) == 0 {
		return "", fmt.Errorf("empty version chain")
	}

	baseVersion := chain[0]
	content, err := encryptionService.DecryptVersionContent(encryption.EncryptedContent{
		Content:   baseVersion.Content,
		ContentIv: baseVersion.ContentIv,
		CipherKey: baseVersion.CipherKey,
		CipherIv:  baseVersion.CipherIv,
	}, symmetricKey)
	if err != nil {
		return "", fmt.Errorf("failed to decrypt base version: %w", err)
	}

	dmp := diffmatchpatch.New()

	for i := 1; i < len(chain); i++ {
		version := chain[i]
		if version.Type != "diff" {
			return "", fmt.Errorf("unexpected version type in chain: %s", version.Type)
		}

		decryptedDiff, err := encryptionService.DecryptVersionContent(encryption.EncryptedContent{
			Content:   version.Content,
			ContentIv: version.ContentIv,
			CipherKey: version.CipherKey,
			CipherIv:  version.CipherIv,
		}, symmetricKey)
		if err != nil {
			return "", fmt.Errorf("failed to decrypt diff version #%d: %w", version.Metadata.VersionNumber, err)
		}

		// Convert delta back to diffs
		diffs, err := dmp.DiffFromDelta(content, decryptedDiff)
		if err != nil {
			return "", fmt.Errorf("failed to parse diff delta for version #%d: %w",
				version.Metadata.VersionNumber, err)
		}

		patches := dmp.PatchMake(content, diffs)
		newContent, applied := dmp.PatchApply(patches, content)

		for _, wasApplied := range applied {
			if !wasApplied {
				return "", fmt.Errorf("failed to apply patches for version #%d",
					version.Metadata.VersionNumber)
			}
		}

		content = newContent
	}

	return content, nil
}
